test/micro/org/openjdk/bench/java/util/regex/Trim.java

```diff
@@ -0,0 +1,177 @@
+ *
+ * some jdk17 numbers:
+ *
+ * Benchmark                    (size)  Mode  Cnt     Score    Error  Units
+ * Trim.find_loop_two_matchers    1024  avgt    8     2.252 ?  0.013  ns/op
+ * Trim.find_loop_usePattern      1024  avgt    8     2.328 ?  0.116  ns/op
+ * Trim.lookBehind_find           1024  avgt    8    21.740 ?  0.040  ns/op
+ * Trim.possessive2_find          1024  avgt    8  7151.592 ? 17.860  ns/op
+ * Trim.possessive2_matches       1024  avgt    8     2.625 ?  0.008  ns/op
+ * Trim.possessive3_find          1024  avgt    8    28.532 ?  1.889  ns/op
+ * Trim.possessive_find           1024  avgt    8  3113.776 ?  9.996  ns/op
+ * Trim.simple_find               1024  avgt    8  4199.480 ? 13.410  ns/op
+ *
+ * TODO: why is simple_find faster than possessive_find, for size below 512 ?
+ *
+ * (cd $(git rev-parse --show-toplevel) && for size in 128 256 512 1024 2048; do make test TEST='micro:java.util.regex.Trim.\\\(simple_find\\\|possessive_find\\\)' MICRO="FORK=2;WARMUP_ITER=1;ITER=4;OPTIONS=-opi $size -p size=$size" |& perl -ne 'print if /^Benchmark/ .. /^Finished running test/'; done)
+ */
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Benchmark)
+public class Trim {
+    /** Run length of non-matching consecutive whitespace chars. */
+    @Param({"16", "64", "256", "1024"})
+    int size;
+
+    /** String containing long interior run of whitespace */
+    public String noMatch;
+
+    public Pattern whitespaceRunPattern;
+    public Pattern eolPattern;
+
+    public Pattern simplePattern;
+    public Pattern possessivePattern;
+    public Pattern possessivePattern2;
+    public Pattern possessivePattern3;
+    public Pattern lookBehindPattern;
+
+    @Setup(Level.Trial)
+    public void setup() {
+        noMatch = "xx" + " \t".repeat(size) + "yy";
+
+        whitespaceRunPattern = Pattern.compile("[ \t]++");
+        eolPattern = Pattern.compile("$", Pattern.MULTILINE);
+
+        simplePattern = Pattern.compile("[ \t]+$");
+        possessivePattern = Pattern.compile("[ \t]++$");
+        possessivePattern2 = Pattern.compile("(.*+[^ \t]|^)([ \t]++)$");
+        possessivePattern3 = Pattern.compile("(?:[^ \t]|^)([ \t]++)$");
+        lookBehindPattern = Pattern.compile("(?<![ \t])[ \t]++$");
+
+        // ad-hoc correctness checking, enabled manually via
+        // make test ...TEST=... MICRO=VM_OPTIONS=-ea;...
+        // TODO: there must be a better way!
+        assert ! simple_find();
+        assert ! possessive_find();
+        assert ! possessive2_find();
+        assert ! possessive2_matches();
+        assert ! possessive3_find();
+        assert ! lookBehind_find();
+        assert ! find_loop_two_matchers();
+        assert ! find_loop_usePattern();
```

Comment:
At a risk of muddling the profiles a bit, this can be rewritten as:

```
 if (!simple_find()) throw new IllegalStateException("simple_find is incorrect");
 ...
```

Reply:
Thanks Aleksey.  But I was hoping for something more magical.

We really want the checking of the result of the benchmark method to be co-located in the source code with the method, but with zero disturbance to the benchmark numbers.  Is such magic possible?

Reply:
I refactored the correctness checking, no longer using the assert facility.